bitcoin/src/internal_macros.rs

```diff
@@ -211,3 +211,108 @@ macro_rules! impl_asref_push_bytes {
+        $(
+            $(#[$($fn_doc_attrs:tt)*])*
+            fn $fn:ident($slf:ident $(, $param_name:ident: $param_type:ty)*) -> $( $ret:ty )? $body:block
+        )*
+    }) => {
+        $(#[$($trait_doc_attrs)*])* $trait_vis trait $trait_name {
+            $(
+                $(#[$($fn_doc_attrs)*])*
+                fn $fn($slf $(, $param_name: $param_type)*) $( -> $ret )?;
+            )*
+        }
+
+        impl $trait_name for $ty {
+            $(
+                fn $fn($slf $(, $param_name: $param_type)*) $( -> $ret )? $body
+            )*
+        }
+    };
+    // With &self, no generics.
+    ($(#[$($trait_doc_attrs:tt)*])* $trait_vis:vis trait $trait_name:ident impl for $ty:ident {
+        $(
+            $(#[$($fn_doc_attrs:tt)*])*
+            fn $fn:ident(&$slf:ident $(, $param_name:ident: $param_type:ty)*) $( -> $ret:ty )? $body:block
+        )*
+    }) => {
+        $(#[$($trait_doc_attrs)*])* $trait_vis trait $trait_name {
+            $(
+                $(#[$($fn_doc_attrs)*])*
+                fn $fn(&$slf $(, $param_name: $param_type)*) $( -> $ret )?;
+            )*
+        }
+
+        impl $trait_name for $ty {
+            $(
+                fn $fn(&$slf $(, $param_name: $param_type)*) $( -> $ret )? $body
+            )*
+        }
+    };
+    // With &self, with generics.
+    ($(#[$($trait_doc_attrs:tt)*])* $trait_vis:vis trait $trait_name:ident impl for $ty:ident {
+        $(
+            $(#[$($fn_doc_attrs:tt)*])*
+            fn $fn:ident$(<$($gen:ident: $gent:ident),+>)?(&$slf:ident $(, $param_name:ident: $param_type:ty)*) $( -> $ret:ty )? $body:block
+        )*
+    }) => {
+        $(#[$($trait_doc_attrs)*])* $trait_vis trait $trait_name {
+            $(
+                $(#[$($fn_doc_attrs)*])*
+                fn $fn$(<$($gen: $gent),+>)?(&$slf $(, $param_name: $param_type)*) $( -> $ret )?;
+            )*
+        }
+
+        impl $trait_name for $ty {
+            $(
+                fn $fn$(<$($gen: $gent),+>)?(&$slf $(, $param_name: $param_type)*) $( -> $ret )? $body
+            )*
+        }
+    };
+    // No self, with generics.
+    ($(#[$($trait_doc_attrs:tt)*])* $trait_vis:vis trait $trait_name:ident impl for $ty:ident {
+        $(
+            $(#[$($fn_doc_attrs:tt)*])*
+            fn $fn:ident$(<$($gen:ident: $gent:ident),+>)?($($param_name:ident: $param_type:ty),*) $( -> $ret:ty )? $body:block
+        )*
+    }) => {
+        $(#[$($trait_doc_attrs)*])* $trait_vis trait $trait_name {
+            $(
+                $(#[$($fn_doc_attrs)*])*
+                fn $fn$(<$($gen: $gent),+>)?($($param_name: $param_type),*) $( -> $ret )?;
+            )*
+        }
+
+        impl $trait_name for $ty {
+            $(
+                fn $fn$(<$($gen: $gent),+>)?($($param_name: $param_type),*) $( -> $ret )? $body
+            )*
+        }
+    };
+    // No self, with generic `<T: AsRef<PushBytes>>`
```

Comment:
Is this even useful? If a function doesn't use `self` it could be just a free-standing function.

Reply:
The function that uses this pattern returns `Self`.

```rust
        fn new_witness_program_unchecked<T: AsRef<PushBytes>>(version: WitnessVersion, program: T) -> Self {
```

Reply:
Derp, we can just return the type directly and not use `Self`, "go directly to jail, do not pass go, do not collect $100" :)

Reply:
Yes, though I now see value in not having to change the code too much.

Reply:
Yes this PR has gotten progressively harder to review everytime I update it. Do you guys want me to split it back up into one patch per extension trait?